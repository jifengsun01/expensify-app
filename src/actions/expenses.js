import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpence = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
        description='', 
        note='', 
        amount=0, 
        createdAt=0
        } = expenseData;
        
        const expense = { description, note, amount, createdAt };
        
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpence({
                id: ref.key,
                ...expense
            }))
        })
    }
}


//REMOVE_EXPENSE
export const removeExpense = ({id}={}) => ({
    type: 'REMOVE_EXPENSE',
    id
})


export const startRemoveExpense = ({id}={}) => {
    return (dispatch) => {
        // Return a promise to allow chaining or handling in the component
        return database.ref(`expenses/${id}`).remove().then(() => {
          // If removal is successful, dispatch the action to update the store
          dispatch(removeExpense({ id }));
        }).catch(error => {
          // Handle error if removal fails (optional)
          console.error('Error removing expense:', error);
        });
    };
}


// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        }); 
    };
};


// SET_EXPENSES

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    })
                });
                dispatch(setExpenses(expenses));
            })
    }
};