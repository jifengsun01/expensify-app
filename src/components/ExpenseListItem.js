import React from "react";
import { Link } from "react-router-dom";
// import { connect } from "react-redux"
// import { removeExpense } from "../actions/expenses";

const ExpenseListItem = ({id, description,amount,createdAt}) => (
    <div>
        <Link to={`/edit/${id}`} >
            <h3>{description}</h3>
        </Link>
        <p> {amount} - {createdAt} </p>
    </div>
);


// const ExpenseListItem = (props) => (
//     <div>
//         <h3>{props.description}</h3>
//         <p>{props.amount} - {props.createdAt}</p>
//     </div>
// );



export default ExpenseListItem