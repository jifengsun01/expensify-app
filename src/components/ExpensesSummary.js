import React from "react";
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = (props) => (
    <div>
        <h1>
            Viewing {props.expenseCount} expense{props.expenseCount !== 1 ? 's' : ''} totaling {numeral(props.expensesTotal / 100).format('$0,0.00')}
        </h1>
    </div>
);

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenses: visibleExpenses,
        expenseCount: visibleExpenses.length, // Calculate expense count based on filtered expenses
        expensesTotal: getExpensesTotal(visibleExpenses) // Calculate total based on filtered expenses
    };
};

export default connect(mapStateToProps)(ExpensesSummary);