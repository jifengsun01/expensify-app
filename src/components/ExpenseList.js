import React from "react";
import {connect} from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expenses) => {
                    return <ExpenseListItem key={expenses.id} {...expenses}/>;
                })
            )
        }
    </div>
);

// const ExpenseList = (props) => (
//     <div>
//         <h1>Expense List</h1>
        
//         {props.expenses.map((expenses) => {
//             return <ExpenseListItem 
//                     key={expenses.id} 
//                     description = {expenses.description}
//                     amount = {expenses.amount}
//                     createdAt = {expenses.createdAt}
//                     />
//         })}
//     </div>
// );



const mapStateToProps = (state) => {
    return {
       expenses: selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);
