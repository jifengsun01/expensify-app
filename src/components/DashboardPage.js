import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpanseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />

    </div>

)

export default ExpanseDashboardPage;
