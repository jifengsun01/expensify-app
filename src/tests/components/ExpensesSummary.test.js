import React from "react";
import { shallow } from 'enzyme';
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with one expense snapshot', () => {
    const wrapper = shallow(
        <ExpensesSummary 
            expenses={expenses[0]} 
            expenseCount={1} 
            expensesTotal={195}
        />
    )
    expect(wrapper).toMatchSnapshot();
}) 

test('should render ExpensesSummary with two expenses snapshot', () => {
    const wrapper = shallow(
        <ExpensesSummary 
            expenses={expenses} 
            expenseCount={3} 
            expensesTotal={24195}
        />
    )
    expect(wrapper).toMatchSnapshot();
}) 

