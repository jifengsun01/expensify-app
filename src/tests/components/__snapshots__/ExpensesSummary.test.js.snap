// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with one expense snapshot 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense
     totaling 
    $1.95
  </h1>
</div>
`;

exports[`should render ExpensesSummary with two expenses snapshot 1`] = `
<div>
  <h1>
    Viewing 
    3
     expense
    s
     totaling 
    $241.95
  </h1>
</div>
`;
